cpu/boards: add xmc1100 cpu & board xmc-2go

sgso opened the pull request:
This adds a new architecture 'xmc1000' with the Infineon XMC1100 Cortex-M0 as the only supported cpu. It is used by the 'xmc-2go' board.

CPU contains drivers for timer, gpio, uart, adc, random, rtc and cpuid. I'll create a SPI one next.

I can provide one or two xmc-2go boards if you want to review.

I'm planning to expand this to the xmc1300 boot kit.

Imports two headers from the upcoming uart overhaul #3265.

sgso commented:
https://github.com/sgso/applications/tree/xmc_test contains a pretty complete test application which can query some of the peripheral subsystems.

thomaseichinger commented:
@sgso Thanks a lot for your contribution. I quickly scanned over it, it looks very good. Will test it in the next days. As you may have seen on the [mailing list](https://lists.riot-os.org/pipermail/devel/2015-September/002982.html) we're currently in feature freeze, so we will probably not be able to merge this within the next week or so. But it will have a high priority right after the release.

haukepetersen commented:
just gave it a quick test - hello world and the default example work fine. Will review in detail once we have the release out (and are out of the feature freeze...). 

Nice job by the way, my previous attempts on porting this platform all failed due to limited time:
https://github.com/haukepetersen/RIOT/tree/add_board_xmc2go
https://github.com/haukepetersen/RIOT/tree/add_board_xmcbootkit

thomaseichinger commented on boards/xmc-2go/Makefile.include:
```diff
@@ -0,0 +1,30 @@
+export CPU_MODEL = xmc1100
+
+# define the default port depending on the host OS
+PORT_LINUX ?= /dev/ttyACM0
```
Please add
```
PORT_DARWIN ?= $(shell ls -1 /dev/tty.usbmodem* | head -n 1)
```

thomaseichinger commented on boards/xmc-2go/Makefile.include:
```diff
@@ -0,0 +1,30 @@
+export FLASHER = $(RIOTBOARD)/$(BOARD)/dist/jlink.sh
+export DEBUGGER = $(RIOTBOARD)/$(BOARD)/dist/jlink.sh
+export DEBUGSERVER = $(RIOTBOARD)/$(BOARD)/dist/jlink.sh
+export RESET = $(RIOTBOARD)/$(BOARD)/dist/jlink.sh
```
would you mind change this to use OpenOCD? XMC* support is added in current version 0.9.0.

sgso replied:
Not at all. But I can't get OpenOCD to play nice with it (won't flash at all, both 0.9.0-rc1 and trunk). Probably made for an older firmware.

thomaseichinger commented on boards/xmc-2go/include/periph_conf.h:
```diff
@@ -0,0 +1,116 @@
+/**
+ * @name Clock system configuration
+ * @{
+ */
```
Documentation artefact?

thomaseichinger commented on cpu/xmc1000/lpm_arch.c:
```diff
@@ -0,0 +1,54 @@
+ */
+
+/**
+ * @ingroup     cpu_stm32f0
```
s/stm32f0/xmc1000/

thomaseichinger commented on cpu/xmc1000/periph/adc.c:
```diff
@@ -0,0 +1,178 @@
+     * Workaround: To enable the analog section of the ADC, write
+     * value 00000001H to register address 40010500H in addition to
+     * the setup as mentioned above.  */
+    *((uint32_t *)0x40010500) = 1UL;
```
Could we use a `#define` for this literal?

thomaseichinger commented on cpu/xmc1000/periph/adc.c:
```diff
@@ -0,0 +1,178 @@
+
+    /* Write value 80008000H to register address 480340E0H and 480340E4H */
+    *((uint32_t *)0x480340E0) = 0x80008000;
+    *((uint32_t *)0x480340E4) = 0x80008000;
```
Also for these.

sgso replied:
Well they don't have names since they're straight from the Errata. I'll think up some.

thomaseichinger commented on cpu/xmc1000/periph/rtc.c:
```diff
@@ -0,0 +1,194 @@
+#if RTC_NUMOF
+
+/* RTC runs of the internal 32.768kHz standby oscillator */
+#define RTC_DIVIDER (0x7fff)
```
Could you move this to `periph_conf.h` please.

thomaseichinger commented on cpu/xmc1000/periph/rtc.c:
```diff
@@ -0,0 +1,194 @@
+static inline void _rtc_disable(void)
+{
+    RTC_WAIT(CTR);
+    RTC->CTR = (RTC_DIVIDER << RTC_CTR_DIV_Pos);
```
Without reading the specs, but wouldn't `RTC->CTR &= ~(1 << RTC_CTR_ENB_Pos)` be less obfuscating?

thomaseichinger commented on cpu/xmc1000/periph/uart.c:
```diff
@@ -0,0 +1,257 @@
+
+    /* "The fractional divider register FDR allows the generation of the
+       internal frequency f_FD, that is derived from the system clock f_PB." */
+    /* uint32_t fdr_step = 590; */
```
Please remove if not needed.

sgso replied:
Don't bother with the uart driver please. It will share the majority of it's code with the SPI one, which already looks a lot better although it's not working yet ;)

thomaseichinger commented on cpu/xmc1000/periph/uart.c:
```diff
@@ -0,0 +1,257 @@
+ */
+static uart_isr_ctx_t uart_ctx[UART_NUMOF];
+
+static inline USIC_CH_TypeDef *_usic(uart_t uart)
```
No blocker, but `inline` defined functions will most certainly not be inlined compiled with `-Os`.

thomaseichinger commented on cpu/xmc1000/periph/uart.c:
```diff
@@ -0,0 +1,257 @@
+    uint32_t fdr_step = 811;
+
+    /* clear register */
+    _usic(uart)->FDR &= ~(USIC_CH_FDR_DM_Msk | USIC_CH_FDR_STEP_Msk);
```
replace be `usic` from line 46?

thomaseichinger replied:
And all following.

thomaseichinger commented:
Awesome work, please see above minor comments.

sgso commented:
rebased & addressed those comments I did reply to.

LudwigOrtmann commented:
Just tested the default example and found it to work =)

sgso commented:
Rebased, adapted to changes in #3265 & cleaning the uart driver.

ArnoRabenstein commented:
Is it possible to add XMC1100 Boot Kit (KIT_XMC11_BOOT_001), which allows to use ARDUINO shields? Maybe it is just a matter of testing, as the designs are very similar except the form factor.
I can provide a few boards if required for testing.

OlegHahm commented:
Btw @thomaseichinger are you happy with the current state of this PR?

thomaseichinger commented:
@OlegHahm I was already before, but I missed the update on the UART driver. Will review again ASAP.

thomaseichinger commented:
@ArnoRabenstein I just tried to flash the mentioned board. Is it pin compatible? I didn't get any UART output.

sgso commented:
@ArnoRabenstein That should be trivial. You could try yourself or contact me on freenode (@sgs) if you want to provide me with a board.

PR is stalled by #3265 anyway. No need to hurry.

sgso commented:
@thomaseichinger They are wired differentrly with UART on P1.3 /1.2

thomaseichinger commented:
@sgso Thank you, will test this tomorrow morning.

sgso commented:
@thomaseichinger Currently the license check is failing because Infineon uses "AS IS" without quotes. Do I change the file or add a pattern to the license check?

thomaseichinger commented:
> Do I change the file or add a pattern to the license check?

@sgso I'd prefer to add a pattern. In case we add more XMCs in the future chances are these are covered then already too.

thomaseichinger commented:
```diff
-.tx = GPIO(P2, 1),     // TX pin
-.rx = GPIO(P2, 2),     // RX pin
+.tx = GPIO(P1, 3),     // TX pin
+.rx = GPIO(P1, 2),     // RX pin
```
Only changing the pins didn't do it for me.
@ArnoRabenstein The UART is connected to the USB socket right?

thomaseichinger commented on boards/xmc-2go/include/periph_conf.h:
```diff
@@ -0,0 +1,112 @@
+        .tx = GPIO(P2, 1),     // TX pin
+        .rx = GPIO(P2, 2),     // RX pin
+        .dsel = 6,             // configure input control register 0
+        .asel = GPIO_ALT_OUT_6 // to use data input DXnG (see page 451)
```
Comment style `//` -> `/*  */`. And could you the section number instead of the page?

ArnoRabenstein replied:
XMC2GO:
  - XMC_TX: P2.1
  - XMC_RX: P2.2

XMC1100 Boot Kit:
  - XMC_TX: P1.2
  - XMC_RX: P1.3

It should work with the following setting:
.usic = USIC0_CH1,   // Usic channel 1
.tx = GPIO(P1,2), // TX Pin of XMC; connected to RX at USB bridge
.rx = GPIO(P1,3), // RX Pin of XMC; connected to TX at USB bridge
.dsel = 0, // USIC0_CH1.DX0A :  P1.3 (Table 14-24, Chapter 14.12)
.asel = GPIO_ALT_OUT_7 // P1.2 : USIC0_CH1.DOUT0 (Table 18-14, Chapter 18.10.3)

thomaseichinger replied:
@ArnoRabenstein Ah, thanks, I confused tx with rx pin. My board has labels `TX>P1.3` and `RX<P1.2`.

ArnoRabenstein replied:
This is Arduino style labeling. The idea is: TX is the sending part of the connected shield or USB Bridge and requires a connection to the input of the microcontroller (XMC), hence TX > P1.3. Anyway, it can be misunderstood easily.
The TX/RX mismatch is a common trouble. With the XMC1000 microcontrollers, the TX and RX function can be swapped by software configuration.

thomaseichinger commented on boards/xmc-2go/include/board.h:
```diff
@@ -0,0 +1,82 @@
+ */
+#define LED_1_ON              (PORT1->OUT |= LED_L_PIN)
+#define LED_1_OFF             (PORT1->OUT &= ~LED_L_PIN)
+#define LED_1_TOGGLE          (PORT1->OUT ^= LED_L_PIN)
```
`LED_L_PIN` is no where defined.

thomaseichinger commented on boards/xmc-2go/include/board.h:
```diff
@@ -0,0 +1,82 @@
+
+#define LED_2_ON              (PORT1->OUT |= LED_R_PIN)
+#define LED_2_OFF             (PORT1->OUT &= ~LED_R_PIN)
+#define LED_2_TOGGLE          (PORT1->OUT ^= LED_R_PIN)
```
`LED_R_PIN` isn't either.

ArnoRabenstein replied:
LED at the left: LED1: P1.0
LED at the right: LED2: P1.1

thomaseichinger commented on cpu/xmc1000/Makefile.include:
```diff
@@ -0,0 +1,7 @@
+
+# only xmc1100 supported for now
+MODEL = $(shell echo $(CPU_MODEL)|tr 'a-z' 'A-Z')
+export CFLAGS += -DCPU_$(MODEL)
```
`-DCPU_MODEL_$(MODEL)` is already defined in `Makefile.include.cortexm_common` and this one seems not to be used.

thomaseichinger commented:
@sgso Do you mind introducing `xmc-common` and `xmc1100-bk` boards? This way we can get in two boards in on PR. 😉

sgso commented:
@thomaseichinger I'll try to separate them. But probably won't find the time today. 

What do you use for flashing? Does openocd work with the xmc1100-bk or did you change the JLINK_OPTS in the Makefile?

thomaseichinger commented:
@sgso Please consider https://github.com/sgso/RIOT/pull/1.

sgso commented:
adapted to #4071

LudwigKnuepfer commented on boards/xmc-1xxx-common/Makefile:
```diff
@@ -0,0 +1,3 @@
+MODULE = $(BOARD)_base
+
+include $(RIOTBASE)/Makefile.base
```
End of file newline missing.

sgso commented:
- XMC1300 MCU & XMC1300 Boot Kit BOARD added (not using any extra features yet)
- added a license pattern `1c-BSD-infineon`
- doxygen fixes, does not introduce warnings anymore.

thomaseichinger commented:
@sgso Great to see even more boards supported! One remark, could you open PRs for each board separately please? (PR for xmc1100-bk and xmc1300-bk depending on this one) This way review process is much easier and more transparent.

haukepetersen commented:
Quick heads up: I opened a new clean PR for the UART changes this week (#4114), would you mind to adapt this PR? And please open separate PRs for each board as @thomaseichinger suggested...

sgso commented:
@haukepetersen I'm on it

haukepetersen commented:
Just noticed one more thing: it seems like both of us wrote a generic script for flashing boards using JLink (mine is included in #3571). If I compared right, our scripts are pretty similar. I would suggest we merge #3571 and adapt this PR to it, would that be ok for you?

OlegHahm commented:
With #4114 being merged the dependency is gone, right?

sgso commented:
@OlegHahm Yes. I still have to adapt it to the now merged UART changes. I'm almost finished with the SPI driver and then I'll rebase, squash, split off the boards and so on.

sgso commented:
@haukepetersen When #3571 gets merged, I'll use it. Can you look into some way to make `make debug` connect to an existing JlinkGDB instance spawned by `make debug-server`? I like to have the server and GDB in separate terminals.

haukepetersen commented:
@sgso #3571 is merged.

About the debug: how so we handle this for openocd based boards? Otherwise I would suggest to add a third debug target like `debug-client` or something like this for all boards? This way you have the shortcut with `make debug` and more control by using the other two targets.

OlegHahm commented:
@sgso, any news?

sgso commented:
@OlegHahm SPI master is working, rebased. Will look into using the FIFOs for uart & spi next or maybe I2C. 

I separated the bootkits from this PR like you all suggested. I also removed the xmc1000-common board part for now because it was kind of empty. It will come back if I need it.

Does anybody want me to separate the xmc-2go also from this PR? (i isolated it into one commit)

haukepetersen commented:
no, please leave the xmc-2go in this PR, makes it easier to test!

haukepetersen commented:
does not build, trouble with static initializers in `spi.c` and `uart.c`:
```
/home/hauke/dev/riot/RIOT/cpu/xmc1000/periph/uart.c: In function `uart_init':
/home/hauke/dev/riot/RIOT/cpu/xmc1000/periph/uart.c:43:5: error: missing braces around initializer [-Werror=missing-braces]
     usic_fdr_t fdr = { 0 };
     ^
/home/hauke/dev/riot/RIOT/cpu/xmc1000/periph/uart.c:43:5: error: (near initialization for `fdr.<anonymous>') [-Werror=missing-braces]
/home/hauke/dev/riot/RIOT/cpu/xmc1000/periph/uart.c:44:5: error: missing braces around initializer [-Werror=missing-braces]
     usic_brg_t brg = { 0 };
     ^
/home/hauke/dev/riot/RIOT/cpu/xmc1000/periph/uart.c:44:5: error: (near initialization for `brg.<anonymous>') [-Werror=missing-braces]
```

haukepetersen commented:
I did a quick, local, hacky fix: the `xmc-2go` works fine for `hello-world` and `default`.

haukepetersen commented on boards/xmc-2go/Makefile.features:
```diff
@@ -0,0 +1,10 @@
+FEATURES_MCU_GROUP = cortex_m0
```
could you please use the same scheme as for e.g. the `boards/stm32f4discovery/Makefie.features` for this file? Would be nice to have them all in the same style...

haukepetersen commented on boards/xmc-2go/board.c:
```diff
@@ -0,0 +1,36 @@
+ */
+
+/**
+ * @defgroup    boards_xmc1000-common Common files for XMC 2Go and XMC 1100 Boot Kit
```
group seems wrong

haukepetersen replied:
EDIT: just remove, you define the group in `board.h` already

haukepetersen commented on boards/xmc-2go/include/periph_conf.h:
```diff
@@ -0,0 +1,133 @@
+ * @name ADC configuration
+ * @{
+ */
+#define ADC_NUMOF          (2U)
```
`ADC_NUMOF` should probably be 1, as you enable only one ADC device, right?

haukepetersen commented on boards/xmc-2go/include/periph_conf.h:
```diff
@@ -0,0 +1,133 @@
+ * @{
+ */
+#define UART_NUMOF         (1U)
+#define UART_0_EN          (1)
```
no need for the `UART_x_EN` defines anymore

haukepetersen commented on boards/xmc-2go/include/periph_conf.h:
```diff
@@ -0,0 +1,133 @@
+ * @name UART configuration
+ * @{
+ */
+#define UART_NUMOF         (1U)
```
better: 
```c
#define UART_NUMOF      (sizeof(uart_instance) / sizeof(uart_instance[0])
```

haukepetersen commented on boards/xmc-2go/include/periph_conf.h:
```diff
@@ -0,0 +1,133 @@
+ * @{
+ */
+#define SPI_NUMOF          (1)
+#define SPI_0_EN           (1)
```
same here

sgso replied:
For SPI we can't use the (sizeof ... / sizeof ...) trick because of the expansion of SPI_NUMOF in `drivers/include/periph/spi.h:37:5`. correct?

haukepetersen commented on cpu/cortexm_common/ldscripts/cortexm_base.ld:
```diff
@@ -105,6 +105,7 @@ SECTIONS
     {
         . = ALIGN(4);
         _srelocate = .;
+        KEEP (*(.veneers .veneers.*));
```
Where is this section coming from? Might make sense to add a comment here

haukepetersen commented on cpu/xmc1000/include/periph/usic.h:
```diff
@@ -0,0 +1,105 @@
+ * @brief   USIC channel initialization function
+ * @{
+ */
+static inline void usic_init(const usic_channel_t *usic_ch,
```
I think it would make sense to de-inline this function for improved code-size. I would suggest to create a `cpu/xmc1000/periph/usic.c` and put the prototype to `cpu/xmc1000/include/periph_cpu.h`

sgso replied:
In theory yes, in practice it is 30692 inlined vs. 30704 linked ;) But I will change this nonetheless. The compiler shall figure it out.

haukepetersen commented on cpu/xmc1000/include/periph/usic_defs.h:
```diff
@@ -0,0 +1,150 @@
+ * @{
+ */
+typedef struct {
+    union {
```
This might be a subjective opinion, but I would prefer not to use bit-addressable register definition. Using these increases code size quite a bit compared to direct access using bitfield defines, e.g. 
```
typedef struct {
    REG32 brg;
    REG32 more_regs...;
} usic_t;

#define USIC_BRG_CLKSEL_POS     (0U)
#define USIC_BRG_CLKSEL(clk)     (clk << USIC_BRG_CLKSEL_POS)
#define USIC_BRG_TMEN      (0x00000008)
...

usic_t *usic;
usic->brg = USIC_BRG_CLKSEL(2) | USIC_BRG_TMEN;
```

sgso replied:
I would like to keep them. I watched the build sizes *very* closely when I introduced them and code size actually decreased slightly while they increase readability a lot. Maybe because they are const and never read out again (I access the uint32_t union only after setting them). If you show me that clang fucks this up, then I'll change it. But gcc is smart enough (at least 5.2).

haukepetersen commented on cpu/xmc1000/ldscripts/xmc1300.ld:
```diff
@@ -0,0 +1,27 @@
+ * @{
+ *
+ * @file
+ * @brief           Memory definitions for the XMC1100
```
s/XMC1100/XMC1300/

haukepetersen commented on cpu/xmc1000/periph/gpio.c:
```diff
@@ -0,0 +1,196 @@
+
+int gpio_init(gpio_t gpio, gpio_dir_t dir, gpio_pp_t pushpull)
+{
+    *(IOCR(gpio)) = (dir | pushpull) << 3;
```
nice!

sgso commented:
@haukepetersen What is your fix for the static initializer problem? And what is the problem anyway? gcc 5.2 (vanilla) does not complain, while 4.9 (codesourcery) does.

sgso commented:
Sorry for the rather intrusive changes to the usic drivers. This enables FIFO use for the UART driver. It makes a big difference. The whole FIFO code path can be disabled by setting UART_ENABLE_FIFOS to 0. This is the current status of things that need to be improved yet:

- [ ] fix isr allocation/use
- [ ] move the poweron()/poweroff() functions into the common usic part (and make them work for SPI)
- [ ] replace my random init hack for the new random seed solution we have now/soon

Incoming:
- [ ] SPI slave mode (almost done)
- [ ] I²C (waiting for some test hardware)

sgso commented:
```
cpu/xmc1000/include/periph/usic.h:167: performance (passedByValue): Function parameter 'brg' should be passed by reference.
cpu/xmc1000/include/periph/usic.h:168: performance (passedByValue): Function parameter 'fdr' should be passed by reference.
```

Can this Travis warning be supressed?

haukepetersen commented:
Just re-compiled, no errors anymore.

Though I see a non-functional issue: Something is screwing up the output of `arm-none-eabi-size`:
```
   text	   data	    bss	    dec	    hex	filename
  33780	    164	  11628	  45572	   b204	home/hauke/dev/riot/RIOT/examples/default/bin/iotlab-m3/default.elf
```
Though a quick look into the `default.map` file reveals, that the data section is actually 448 (0x1c0) byte:
```
.relocate       0x20000000      0x1c0 load address 0x100040d8
```
I still don't understand what your newly introduced `.veneers` section does, but this seems to cause it.

sgso commented:
@haukepetersen The XMC1000 does a weird thing with its ISR vectors: "The vector table is remapped to locations on SRAM, with each vector allocated to a size of 4 bytes.
Typically, the allocated size is not sufficient for an exception or interrupt. Therefore, a branch
instruction is needed to jump to the actual handler at another location. This is achieved via veneers.
Basically, a veneer acts as an intermediate target of the instruction and then sets the PC to the actual location."

That's what the section is for. And that's why I have to use asm inside `vectors.c`.

haukepetersen commented:
Hm, I see, but seems overly complicated to me (although not surprising when looking at other Infineon MCUs). One thing that I just don't really get into my head: The xmc is using a Cortex-M0 core. Now the core for the cortex-m MCUs includes the interrupt controller (NVIC). This would lead me to the assumption, that the interrupt handling would be the same over all Cortex-M based CPUs?!

The reason why I press this so much is to make very very sure, that we not just using some exotic interrupt handling for the XMC CPUs because of Infineon example/periph lib code, but that it really MUST be done like this...

sgso commented:
@haukepetersen: It must be done like that. I think the way it works is that the XMCs have this rather large boot ROM code called SSW (Startup Software) which you cannot circumvent and the interrupt table is at its usual place but in a region where you cannot write to. So they get redirected. 
This manual explains most of the startup sequence: http://www.infineon.com/dgdl/DEV_XMC1000_C-Start_v1_0.pdf?fileId=db3a30433e78ea82013e7e060d3a056c&ack=t


Maybe @ArnoRabenstein can say something about this?

haukepetersen commented:
Hmm, I see... I guess I don't have to like it, though :-(

Then I would propose to put the `.veneers` stuff in it's own section as this:
```


    .relocate : AT (_etext)
    {
        . = ALIGN(4);
        _srelocate = .;
        *(.ramfunc .ramfunc.*);
        *(.data .data.*);
        . = ALIGN(4);
        _erelocate = .;
    } > ram

    /* The XMC1000 family needs this for their vectors to work.
     * The segment is set up in cpu/xmc1000/vectors.c */
    .veneers :
    {
        . = ALIGN(4);
        KEEP (*(.veneers .veneers.*));
        . = ALIGN(4);
    } > ram
```
so at least the `.relocate` section, aka `data`, is recognized correctly by `size`, although the RAM space taken by the `veneer` stuff is not shown... I am just unsure about the alignment, but aligning on board boundaries should not hurt here?!

sgso commented:
@haukepetersen: How about I offset the RAM section by the amount of memory needed for the table (0x1c0, 448 Bytes) and then just relocate the veneers from a simple const variable in .text to RAM in a pre_startup()? It would not need any change to the cortexm_common linker script and would not mess with objsize.

sgso commented:
Latest push adds I2C driver.

@thomaseichinger Not so happy about that WIP label. It works, is stable, well documented and the points in progress is largely stuff that other boards don't even bother to implement (right).

thomaseichinger commented:
@sgso The label was not meant to offend you. But this PR seems to be in progress as you mentioned to add features and they are on a regular basis. It does not target stability or quality.

sgso commented:
@thomaseichinger Ok (not offended).

OlegHahm commented:
Why did you close?

sgso commented:
@OlegHahm For  one thing I was fed up with rebasing this for the nth time. It was in a perfectably mergeable state at _several_ points in time. It was in a better shape than half of the other platforms that I looked into while developing it (in terms of clearness, documentation, feature set, use of modern riot functionality, code size, etc. pp -- hint: there's some bit rot). But instead of any kind of helpful code review/critique, I got burdened with a sequence for contradicting requests (merge this in vs. factor this out then suddenly separated commits were ok). I have better things to do with my _free_ time, especially since I'm working on my bachelor thesis now (which I wanted to write in the embedded field initially but this whole story turned me off, sorry).

To be clear about this: Even now, I _would_ find the hours to rebase this, but in the past it always felt like an enormous waste of time. This work will resurface in my repository when I'm done with my BA and live an isolated life there, unhassled by the development/contribution processes of this project, which I think are flawed.

OlegHahm commented:
@sgso, sorry to hear that. Reading through the discussion in this PR again, I'm still unable to find any of these problems being expressed earlier. If you were and are unhappy with the review process here, it might be helpful to state this before you get too frustrated to continue. Also, it would appreciated if you have any suggestions to improve the review process.

(Btw. most people are doing the reviews in their free time, too.)
